samples/connectivity/callnotification/src/main/java/com/example/platform/connectivity/callnotification/NotificationSource.kt

```diff
@@ -0,0 +1,203 @@
+import androidx.core.app.NotificationCompat.PRIORITY_MAX
+import androidx.core.app.NotificationCompat.VISIBILITY_PUBLIC
+import androidx.core.app.Person
+import androidx.core.content.getSystemService
+
+/**
+ * Source for creating notifications and notification channels
+ */
+class NotificationSource<T>(
+    private val context: Context,
+    private val notificationBroadcastClass: Class<T>,
+) {
+
+    companion object {
+
+        const val ChannelId = "1234"
+
+        /*
+        Notification state sent via in intent to inform receiving broadcast what action the user wants to take
+         */
+        enum class NotificationState(val value: Int) {
+            ANSWER(0),
+            REJECT(1),
+            CANCEL(2),
+            CLEARED(3);
+
+            companion object {
+                fun valueOf(value: Int) = NotificationState.values().find { it.value == value }
+            }
+        }
+
+        //Get ringtone location
+        private val ringToneUri = RingtoneManager.getDefaultUri(RingtoneManager.TYPE_RINGTONE)
+        const val NOTIFICATION_ACTION = "NOTIFICATION_ACTION"
+
+        //Channel for on going call, this has low importance so that the notification is not always shown
+        @SuppressLint("NewApi")
+        val notificationChannelOngoing = NotificationChannel(
+            ChannelId, "Call Notifications",
+            NotificationManager.IMPORTANCE_LOW,
+        )
+
+        //Notification channel for incoming calls. Will play ringtone, will be no dismissible and will stay on screen until user interacts with notification.
+        @SuppressLint("NewApi")
+        fun notificationChannelIncoming(): NotificationChannel {
+            val notification = NotificationChannel(
+                ChannelId, "Call Notifications",
+                NotificationManager.IMPORTANCE_HIGH,
+            )
+            notification.importance = NotificationManager.IMPORTANCE_HIGH
+            notification.setSound(
+                ringToneUri,
+                AudioAttributes.Builder()
+                    .setContentType(AudioAttributes.CONTENT_TYPE_SONIFICATION)
+                    .setLegacyStreamType(AudioManager.STREAM_RING)
+                    .setUsage(AudioAttributes.USAGE_NOTIFICATION_RINGTONE)
+                    .build(),
+            )
+
+            return notification
+        }
+
+        @SuppressLint("NewApi")
+        var fakeCaller = Person.Builder()
+            .setName("Jane Doe")
+            .setImportant(true)
+            .build()
+    }
+
+    private val notificationManager = context.getSystemService<NotificationManager>()!!
+
+    //Intent and pending intent for full screen activity.
+    private val intent = Intent()
+        .setAction(Intent.ACTION_MAIN)
+        .setPackage(context.packageName)
+
+    private val pendingIntent =
+        PendingIntent.getActivity(context, 0, intent, PendingIntent.FLAG_IMMUTABLE)
+
+    init {
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
```

Comment:
I think with NotificationCompat class you can avoid doing the OS check yourself

Reply:
Channels are only required for Android 8 and above and setting priority in the notification covers devices using 7.1 and below.

Reply:
Channels are only required for O and above

Reply:
What I mean is that you can use NotificationManagerCompat instead to avoid this checks:
https://developer.android.com/reference/androidx/core/app/NotificationManagerCompat

```
val managerCompat = NotificationManagerCompat.from(requireContext())
managerCompat.createNotificationChannels(listOf( notificationChannelIncoming(), notificationChannelOngoing))
```

Reply:
Great, thanks for the info. 

Did not know this.